src/Xml/Dom/Traverser/Visitor/RemoveNamespaces.php

```diff
@@ -80,11 +81,9 @@ public function onNodeLeave(DOMNode $node): Action
             $namespaces = $namespaces->filter($this->filter);
         }
 
+        // TODO : convert to is_xmlns_attribute based visitor instead of acting on element level
         foreach ($namespaces as $namespace) {
-            $node->removeAttributeNS(
-                $namespace->namespaceURI,
-                $namespace->prefix
-            );
+            $node->removeAttributeNode($namespace);
```

Comment:
This doesn't seem to work as expected anymore (in comparison to legacy dom):

```php
$xml = \DOM\XMLDocument::createFromString('
        <hello xmlns:a="http/a" xmlns:z="http/z" version="1.9" target="universe">
            <item xmlns="http://raw" id="1" sku="jos">Jos</item>
            <x:item xmlns:x="http://x" sku="jaak" id="2">Jaak</x:item>
        </hello>
');

$hello = $xml->documentElement;
$item1 = $hello->firstElementChild;
$item2 = $hello->lastElementChild;

$item1->removeAttributeNode($item1->getAttributeNode('xmlns'));
$item2->removeAttributeNode($item2->getAttributeNode('xmlns:x'));
$hello->removeAttributeNode($hello->getAttributeNode('xmlns:a'));
$hello->removeAttributeNode($hello->getAttributeNode('xmlns:z'));

echo $xml->saveXML();
```

Output:

```xml
<?xml version="1.0" encoding="UTF-8"?>
<hello version="1.9" target="universe">
    <item xmlns="http://raw" id="1" sku="jos">Jos</item>
    <x:item xmlns:x="http://x" sku="jaak" id="2">Jaak</x:item>
</hello>
```

output in legacy:

```xml
        <hello version="1.9" target="universe">
            <item id="1" sku="jos">Jos</item>
            <item sku="jaak" id="2">Jaak</item>
        </hello>
```

The expectation is that all xmlns declarations are gone - including prefixes.

Reply:
The new output is correct because of the same reason as in https://github.com/veewee/xml/pull/74#discussion_r1500981687, you can't really change the namespace after an element was created.

Porting to JavaScript to double check, and I get the same output:
```js
xml = (new DOMParser).parseFromString(`
        <hello xmlns:a="http/a" xmlns:z="http/z" version="1.9" target="universe">
            <item xmlns="http://raw" id="1" sku="jos">Jos</item>
            <x:item xmlns:x="http://x" sku="jaak" id="2">Jaak</x:item>
        </hello>
`, 'text/xml');

hello = xml.documentElement;
item1 = hello.firstElementChild;
item2 = hello.lastElementChild;

item1.removeAttributeNode(item1.getAttributeNode('xmlns'));
item2.removeAttributeNode(item2.getAttributeNode('xmlns:x'));
hello.removeAttributeNode(hello.getAttributeNode('xmlns:a'));
hello.removeAttributeNode(hello.getAttributeNode('xmlns:z'));

(new XMLSerializer).serializeToString(xml);
```

Reply:
> and why this would be so hard to do

Not really hard to implement ourselves, with the new XML serializer in place and the custom namespace support code this is doable.

> Maybe adding a few methods to make renaming /removing easier in general might make sense?

I think that makes sense and I can include that in my next RFC.
If you want, you can give a list of namespace manipulation methods that you'd need, and we can discuss them and I can check how difficult it would be to implement them.

One thing I am worried about a bit though is potential API clashes. For example, let's say we add an API `setNamespace` then it could become a problem if the DOM spec were to add a method with the same name that's different than what we have. I'd have to think how we could mitigate that (e.g. prefixing those method names could be one way, but is a bit ugly).
As always, I'm open for suggestions and ideas :slightly_smiling_face:

Reply:
The alternative is to recreate the element, and then move the attributes and child nodes from the old to the new element. But that's cumbersome.

Reply:
> If you want, you can give a list of namespace manipulation methods that you'd need, and we can discuss them and I can check how difficult it would be to implement them.

This library supports following `renames`:
https://github.com/veewee/xml/blob/3.x/tests/Xml/Dom/Manipulator/Node/RenameTest.php
(Every test contains a specific rename case you could use as an example)

This library supports following `removes`:
https://github.com/veewee/xml/blob/3.x/tests/Xml/Dom/Manipulator/Node/RemoveTest.php
(Every test contains a specific remove case you could use as an example)

I believe the remove part already works for most types of nodes, yet removing namespaces is not fully possible yet as you can see in the example above. 


> One thing I am worried about a bit though is potential API clashes. [...] I'd have to think how we could mitigate that (e.g. prefixing those method names could be one way, but is a bit ugly).

If we go for very straight-forward things like rename and remove, I don't think there will be big clashes. If they decide to specify it at spec level, it will probably work very similar to what you are building. I wouldn't go for something like `setNamespace`, cause that might be a bit weird. If you really want to make sure the DOM classes are according to the spec, another option might be to introduce some helper classes. For example : `(new ExtendedNodeManipulator($node))->rename(....)`. But that might become a bit confusing for the end users.


> The alternative is to recreate the element, and then move the attributes and child nodes from the old to the new element. But that's cumbersome.

It's possible to recreate indeed. But some actions are going very far: for example renaming an xmlns attribute requires to do a lot of manipulations at a lot of layers. So it wouldn't "simply" be renaming.

Reply:
I've spent some time thinking on this, both on how this would work for PHP developers, and how this would work internally in the extension.

First, let's reiterate:

Removing an xmlns attribute is already possible, yet it does not change the element / attribute namespace information because xmlns attributes are merely "hints". It's even possible to have an element in a namespace without a xmlns attribute declaring that namespace in the document. The serializer adds it in this case to make the document parse correctly.

If we were to introduce manipulation methods that work directly on xmlns attributes, then we probably create some confusion (e.g. why does `$xmlns_attr->value = ...` not change the namespaceURI of elements whereas `$xmlns_attr->manipulateURI()` work?). Also, if the attribute doesn't exist in the document then you can't use such an API.

Instead, what if we add a method for DOM\Element and DOM\Attr called `renameNode`?
Signature: `function renameNode(?string $namespace, string $qualifiedName)`. Inspired by the scrapped method from DOM3: https://www.w3.org/TR/2004/REC-DOM-Level-3-Core-20040407/core.html#Document3-renameNode
This allows:
* Changing the namespace uri of the node using the first argument. You can even put an element out of a namespace by passing "" or NULL, effectively removing the namespace from an element.
* Changing the prefix.
* Changing the local name.

Unlike old DOM, it would actually change the namespace information of those nodes (and thus change $node->prefix & $node->namespaceURI) and leave any xmlns attributes intact (as the xmlns attributes and node namespace information are not actually "linked" to each other).

I guess what may be missing for you here is being able to change the namespace of all children in the same namespace with a single method call?

Reply:
> I guess what may be missing for you here is being able to change the namespace of all children in the same namespace with a single method call?

It's kind of why I suggested a setNamespace method, it could take a $deep argument.
But the API design of that is not really optimal.

Reply:
Making it possible to rename elements and attributes seems like a no-brainer and sounds good to do.

So maybe lets zoom in again on xmlns changes: I think the use-cases for this are rather limited.
In this library, I am providing:

* A way to optimize namespaces : It lists all available namespaces in the document, adds them (sorted) to the root namespace with a prefixes name (for example ns1, ns2, ns3, ...) and makes sure all nodes use this namespace. It's based on https://genericparallel.com/blog/2018/02/optimize-namespaces-to-reduce-xml-size/. Now I can imagina it makes sense to have something like `XMLDocument->optimizeNamespaces()` so that it can be used by anyone. This is basically my only use case for renaming xmlns at the moment I think. It is also the heart of the `comparable()` configuratort which : canonicalizes, optimizes namespaces and sorts attributes to make the XML fully comparable with another comparable XML. So one step further, it might also make sense to add this one directly to `XMLDocument` so that tools like phpunit etc can use this to compare XML strings.

* This package provides a traverser visitor for removing namespaces : https://github.com/veewee/xml/blob/3.x/src/Xml/Dom/Traverser/Visitor/RemoveNamespaces.php. It is mostly used in combination with the xml_decoder that transforms an XML document into a structured array. There are options to : remove specific prefix, remove prefix based on namespaceURI, remove unprefixed/prefixed namespaces, ... This one is the main use-case for removing nested namespaces from XML so that you can use the resulting array as-if there were no XML namespaces inside the document or to run easier xpath queries, ...

* The sort attributes logic also removes xmlns attributes, but then re-adds it again to make sure the attributes are sorted the right way. For this one, removing the childrens namespaces would not be necasarry, since it's getting readded again later.

I think those 3 are my main use-cases which need to be covered somehow.
If you think it would be easier to keep the logic as-is and provide other means to do the same actions, that might  be OK as well. WDYT?

Reply:
First of all, I could be wrong with what I'm going to answer, as I haven't yet tried these things. And I'm not sure when I can try adding the new APIs because it depends on how much free time I have.
But here goes nothing :-)

## Optimizing namespaces

I'm not sure I'd like to add an `optimizeNamespaces()` method to `XMLDocument` for 2 reasons: it's already a quite specific operation, and it's not in a spec somewhere. I think it's better to provide lower level methods that can be used to build higher level ones such as `optimizeNamespaces()`.

As a side note, depending on the specific definition of comparable, you can also use `DOMNode::isEqualNode` that was added in 8.3, although it seems you want to disregard the prefixes in which case that method won't work for you.

I think you can implement the `optimizeNamespaces()` method using the proposed `renameNode(?string $namespace, string $qualifiedName)` method though. The general idea:
1. Remove all xmlns attributes.
2. Traverse the document and give a unique prefix to each (prefix, uri) pair. Call `$node->renameNode($node->namespaceURI, $unique_prefix_for_this_uri . ':' . $node->localName);`. This will keep the namespace uri of the node, but change its prefix.
3. Create the xmlns attributes for each unique id.

So I think the `renameNode` is quite versatile and sufficient. WDYT? I didn't try this because I didn't implement `renameNode` yet.

## Removing namespaces

I think again that removing a namespace from a node can be done with `renameNode`, e.g. `$node->renameNode(NULL, $node->localName)`. This should allow you to remove specific prefixes, or based on namespaceURI and prefixed/unprefixed namespaces. You would need to traverse the document subtree yourself though. After you did all the renaming, you can remove the xmlns attributes as they'd serve no purpose anymore.

Sadly we don't really have a convenient tree traverser in ext-dom (yet), so doing the traversal manually is a bit of a pain. Although I could implement https://dom.spec.whatwg.org/#interface-treewalker

## Sorting

The serializer adds namespace declarations in the output in the order in which elements or attributes reference a specific namespace, so we can't rely on that sadly.
What you could do though (I think) is:
1. Traverse a subtree for all namespace usages, and make a sorted array of their names
2. Unlink all xmlns attributes
3. Manually add the necessary xmlns attributes, in the sorted order

Reply:
It all sounds logical and reasonable. 
I think we can agree there is a need for `renameNode` first so that I can try these concepts out.
As long as it does not exist, I won't be able to really verify your train of thoughts.

> And I'm not sure when I can try adding the new APIs because it depends on how much free time I have.

Take all the time you need. There are more important things in life :)


> I'm not sure I'd like to add an optimizeNamespaces() method to XMLDocument for 2 reasons: it's already a quite specific operation, and it's not in a spec somewhere. I think it's better to provide lower level methods that can be used to build higher level ones such as optimizeNamespaces().

Thats true, but because it is a very specific operation - there is little room for alternate implementations there. Now to be clear : I don't expect you to implement this. I just wanted to share the use-cases and possible options. So I'm fine with either having it through low-level or dedicated APIs.


> Sadly we don't really have a convenient tree traverser in ext-dom (yet), so doing the traversal manually is a bit of a pain. Although I could implement https://dom.spec.whatwg.org/#interface-treewalker

Didn't know that existed in the spec. I've provided a custom implementation to do these kind of opertations: https://github.com/veewee/xml/blob/3.x/src/Xml/Dom/Traverser/Traverser.php
Having a dedicated API might be nice though.


> The serializer adds namespace declarations in the output in the order in which elements or attributes reference a specific namespace, so we can't rely on that sadly.

I'm a bit confused by this explanation. However, previous implementation didn't even sort namespaces - so I might just exclude that again. Once namespace optimization is there, that system already deals with the namespace sorts (or at least in previous legacy implementation). I'm sure we can figure something out here once the low-level APIs are available.

Reply:
> I just wanted to share the use-cases and possible options.

And I'm glad for having this discussion about the use-cases, it's insightful and helps to see where the pain points are :slightly_smiling_face: 

> So I'm fine with either having it through low-level or dedicated APIs.

:+1: 

> I'm a bit confused by this explanation. However, previous implementation didn't even sort namespaces - so I might just exclude that again. Once namespace optimization is there, that system already deals with the namespace sorts (or at least in previous legacy implementation). I'm sure we can figure something out here once the low-level APIs are available.

Yeah, it was written in an unclear way. What I meant was that it's not possible to use tricks/hacks to force the serializer to output them in a sorted order. So you'd have to add all xmlns attributes that are necessary for serialization yourself such that you have full control and such that the serializer won't add "missing" ones.

That being said, looks like the spec-compliance RFC got accepted. So I'm going to spend time now on getting that reviewed and merged, and then I'll go back to this and try to get the missing APIs implemented. Once I got an implementation that works I'll ping you again.

Thanks for the discussion!

Reply:
Cool, congrats with the acceptation of the PR!
It's my pleasure :)